### PR TITLE
Ip Cogniton Auth middleware

### PR DESCRIPTION
**Pull request by sashah2**
# Summary
Implement Ip Cognito Auth middleware for consumer services. 

The middleware allow to get an access to an action if an IP is in whitelist of Cognito JWT token is valid.

### CODE REVIEW

File: httpmw/Ip_cognito_auth.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+					return
+				}
+			}
+		}
+
+		realm := "Authorization Required"
+		authHeader := strings.Split(c.GetHeader("Authorization"), "Bearer ")
+
+		if len(authHeader) != 2 || authHeader[1] == "" {
+			c.Header("WWW-Authenticate", realm)
```
Comment (protsack-stephan) [inline httpmw/Ip_cognito_auth.go]: Do we really need `WWW-Authenticate` header here?

File: httpmw/Ip_cognito_auth.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			c.AbortWithStatus(http.StatusUnauthorized)
+			return
+		}
+
+		token := authHeader[1]
+		res, err := svc.GetUser(&cognitoidentityprovider.GetUserInput{AccessToken: &token})
+
+		if err != nil {
+			log.Println(err)
+			c.Header("WWW-Authenticate", realm)
```
Comment (protsack-stephan) [inline httpmw/Ip_cognito_auth.go]: Do we really need `WWW-Authenticate` header here?

Comment (protsack-stephan) [inline httpmw/Ip_cognito_auth.go]: Do we really need `WWW-Authenticate` header here?

File: httpmw/Ip_cognito_auth.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+				if checkIP(ipRange, c.ClientIP()) {
+					return
+				}
+			}
+		}
+
+		realm := "Authorization Required"
+		authHeader := strings.Split(c.GetHeader("Authorization"), "Bearer ")
+
+		if len(authHeader) != 2 || authHeader[1] == "" {
```
Comment (protsack-stephan) [inline httpmw/Ip_cognito_auth.go]: This is a potential panic in case of empty `Authorization` header.

File: httpmw/Ip_cognito_auth.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		if len(ipRanges) > 0 {
+			for _, ipRange := range ipRanges {
+				if checkIP(ipRange, c.ClientIP()) {
+					return
+				}
+			}
+		}
+
+		realm := "Authorization Required"
+		authHeader := strings.Split(c.GetHeader("Authorization"), "Bearer ")
```
Comment (protsack-stephan) [inline httpmw/Ip_cognito_auth.go]: I suggest replacing this with:
```go
token := strings.Replace(c.GetHeader("Authorization"), "Bearer ", "", 1)

if len(token) <= 0 {
	httperr.Unauthorized(c)
	c.Abort()
	return
}
```

File: httpmw/Ip_cognito_auth.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+			return
+		}
+
+		token := authHeader[1]
+		res, err := svc.GetUser(&cognitoidentityprovider.GetUserInput{AccessToken: &token})
+
+		if err != nil {
+			log.Println(err)
+			c.Header("WWW-Authenticate", realm)
+			c.AbortWithStatus(http.StatusUnauthorized)
```
Comment (protsack-stephan) [inline httpmw/Ip_cognito_auth.go]: I suggest we replace this with:
```go
httperr.Unauthorized(c)
c.Abort()
```

**Review by protsack-stephan — COMMENTED** [review]
LGTM! Couple small comments.

File: httpmw/Ip_cognito_auth.go (hunk 2 above)
Comment (sashah2) [reply httpmw/Ip_cognito_auth.go]: Just as we have in other middlewares

Comment (sashah2) [reply httpmw/Ip_cognito_auth.go]: removed

File: httpmw/Ip_cognito_auth.go (hunk 3 above)
Comment (sashah2) [reply httpmw/Ip_cognito_auth.go]: Done

**Comment by sashah2** [comment]
@protsack-stephan thank you for the review. All changes have been made

**Pull request merged**